src/app/tests/TestInteractionModel.cpp

```diff
@@ -0,0 +1,193 @@
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file implements a test for  CHIP Interaction Model Message Def
+ *
+ */
+
+#include <app/InteractionModelEngine.h>
+#include <core/CHIPCore.h>
+#include <core/CHIPTLV.h>
+#include <core/CHIPTLVDebug.hpp>
+#include <core/CHIPTLVUtilities.hpp>
+#include <messaging/ExchangeContext.h>
+#include <messaging/ExchangeMgr.h>
+#include <messaging/Flags.h>
+#include <platform/CHIPDeviceLayer.h>
+#include <support/ErrorStr.h>
+#include <support/UnitTestRegistration.h>
+#include <system/SystemPacketBuffer.h>
+#include <system/TLVPacketBufferBackingStore.h>
+#include <transport/PASESession.h>
+#include <transport/SecureSessionMgr.h>
+#include <transport/raw/UDP.h>
+
+#include <nlunit-test.h>
+chip::SecureSessionMgr gSessionManager;
+chip::Messaging::ExchangeManager gExchangeManager;
+chip::TransportMgr<chip::Transport::UDP> gTransportManager;
+const chip::Transport::AdminId gAdminId = 0;
+
+namespace chip {
+namespace app {
+class TestInteractionModel
+{
+public:
+    static void TestReadClient(nlTestSuite * apSuite, void * apContext);
+    static void TestReadHandler(nlTestSuite * apSuite, void * apContext);
+
+private:
+    static void GenerateReportData(nlTestSuite * apSuite, void * apContext, System::PacketBufferHandle && aPayload);
+};
+
+void TestInteractionModel::GenerateReportData(nlTestSuite * apSuite, void * apContext, System::PacketBufferHandle && aPayload)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    chip::System::PacketBufferTLVWriter writer;
+    writer.Init(std::move(aPayload));
+
+    ReportData::Builder reportDataBuilder;
+
+    err = reportDataBuilder.Init(&writer);
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+    reportDataBuilder.SuppressResponse(true);
+    NL_TEST_ASSERT(apSuite, reportDataBuilder.GetError() == CHIP_NO_ERROR);
+
+    reportDataBuilder.MoreChunkedMessages(false);
+    NL_TEST_ASSERT(apSuite, reportDataBuilder.GetError() == CHIP_NO_ERROR);
+
+    reportDataBuilder.EndOfReportData();
+    NL_TEST_ASSERT(apSuite, reportDataBuilder.GetError() == CHIP_NO_ERROR);
+
+    err = writer.Finalize(&aPayload);
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+}
+
+void TestInteractionModel::TestReadClient(nlTestSuite * apSuite, void * apContext)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    chip::app::ReadClient readClient;
+
+    chip::System::PacketBufferHandle buf = chip::System::PacketBufferHandle::New(chip::System::PacketBuffer::kMaxSize);
+    err                                  = readClient.Init(&gExchangeManager, nullptr);
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+    err = readClient.SendReadRequest(chip::kTestDeviceNodeId, gAdminId);
+    NL_TEST_ASSERT(apSuite, err == CHIP_ERROR_NOT_CONNECTED);
+
+    GenerateReportData(apSuite, apContext, buf.Retain());
+
+    err = readClient.ProcessReportData(std::move(buf));
```

Comment:
Shouldn't we call into the public API that exchange context would call into, which would also let us test the "did I get the right response?" handling and so on.  Similar elsewhere in this file.

Reply:
the public API OnMessageReceived don't have returned error code for check, OnMessageReceived is virtual function in Messaging::ExchangeDelegate, so i test internal function ProcessReportData

Reply:
I made a suggestion for handling this in the review I just did.